### PR TITLE
8257069: C2: Clarify and sanity test RegMask/RegMaskIterator

### PR DESCRIPTION
**Pull request by cl4es**
This patch adds a sanity test to RegMask. It's not (yet) exhaustive, but it covers most common operations and iteration.

While implementing this test I noticed that the code have a strong, implied assumption that RM_SIZE is even on 64-bit platforms, and a few things could have broken badly if RM_SIZE was odd (mismatch between set_AllStack/is_AllStack, extra bits after CHUNK_SIZE). That RM_SIZE is even is thankfully an invariant, since the AD preprocessor aligns up RM_SIZE. Add a static assert to this effect, and prefer _RM_SIZE for clarity.

The iteration algorithm in RegMaskIterator, which I borrowed from IndexSetIterator, has garnered a few raised eyebrows. I think the quirky scheme of not shifting out the bit of interest but instead clear is mainly explained by the need to avoid undefined behavior in the corner case when only the highest bit is set. I've added some commentary to try and clarify the quirks.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux additional | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (8/8 passed) | ✔️ (2/2 passed) | ❌ (2/2 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ❌ (1/9 failed) |    |  ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x64 (hs/tier1 runtime)](https://github.com/cl4es/jdk/runs/1456135012)
- [Linux x86 (build debug)](https://github.com/cl4es/jdk/runs/1455662478)
- [Linux x86 (build release)](https://github.com/cl4es/jdk/runs/1455662467)

### Issue
 * [JDK-8257069](https://bugs.openjdk.java.net/browse/JDK-8257069): C2: Clarify and sanity test RegMask/RegMaskIterator

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer) ⚠️ Review applies to 8c226d22ced5784165478e4238ea9b28daedf282
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1437/head:pull/1437`
`$ git checkout pull/1437`

### CODE REVIEW

**Review by JornVernee — APPROVED** [review]
Okay-ing the changes in RegMaskIterator, since we've had some offline discussion about it already :)

File: src/hotspot/share/opto/regmask.hpp [hunk 1]
```diff
@@ -158,9 +161,13 @@ class RegMask {
...
   // unbounded in size.  Returns FALSE if mask is finite size.
-  bool is_AllStack() const { return _RM_UP[_RM_SIZE - 1U] >> (_WordBits - 1U); }
+  bool is_AllStack() const {
+    return (_RM_UP[_RM_SIZE - 1U] & (uintptr_t(1) << (_WordBits - 1U))) != 0;
+  }
 
-  void set_AllStack() { Insert(OptoReg::Name(CHUNK_SIZE-1)); }
+  void set_AllStack() {
+    _RM_UP[_RM_SIZE - 1U] |= (uintptr_t(1) << (_WordBits - 1U));
+  }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/regmask.hpp:170]: `_WordBits - 1U` is used in several places in this file. Should we add and use new _WordBitsM1 value instead?

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by rose00** [comment]
This is not a review comment, because you didn't change the algorithm, but:  That code is hard to follow, equally before and after.  There is a lot of delicate "off-by-one" logic in it.  Since it Just Words ™ we might choose to ignore this issue, but it is tech. debt that must be paid whenever we might try to improve the code in the future.

More detail, FWIW:

The treatment of _current_word as a state variable is very confusing, because it is a mish-mash of two styles of a bitmask state.  Style one is “keep the bits in place but clear them one by one”.  Style two is “shift out the bit positions one by one.”  The style here is “shift out all the bits except the current one, but clear it for good measure”.  That’s IMO hard to read and reason about.  If I were writing it from scratch, I would prefer the first style, where there is no shifting of _current_word, just bit-by-bit clearing to zero.  This does require a second _reg-like state variable to restart the bit-search, or else a constant-time FFS operation.  Failing that, I’d prefer the second style, shifting out not only the zero positions, but the next one position.  But that’s probably almost as liable to O-B-O errors, since you have to shift >> (next_bit+1), and increment _reg by a similar amount.  Anyway, surely there’s a way to make code like this more self-explanatory.  Comments would be a kind service to future readers.

As a variation of the "whiteboard effect" (which is what happens when you explain something to a colleague and at the same time get better understanding of what you are explaining), adding well-written comments to code like this often suggests a better way to formulate the code itself.  At least, that's been my experience.
